packages/editor-ui/src/components/InlineExpressionEditor/InlineExpressionTip.vue

```diff
@@ -1,127 +1,185 @@
+import { datatypeCompletions } from '@/plugins/codemirror/completions/datatype.completions';
+import { watchDebounced } from '@vueuse/core';
+import { FIELDS_SECTION } from '@/plugins/codemirror/completions/constants';
+import { isCompletionSection } from '@/plugins/codemirror/completions/utils';
+
+type TipId = 'executePrevious' | 'drag' | 'default' | 'dotObject' | 'dotPrimitive';
+
+type Props = {
+	editorState?: EditorState;
+	unresolvedExpression?: string;
+	selection?: SelectionRange;
+};
+
+const props = withDefaults(defineProps<Props>(), {
+	editorState: undefined,
+	unresolvedExpression: '',
+	selection: () => EditorSelection.cursor(0),
+});
 
 const i18n = useI18n();
 const ndvStore = useNDVStore();
 
-const props = defineProps<{ tip?: 'drag' | 'default' }>();
+const canAddDotToExpression = ref(false);
+const resolvedExpressionHasFields = ref(false);
+
+const canDragToFocusedInput = computed(
+	() => !ndvStore.isDNVDataEmpty('input') && ndvStore.focusedMappableInput,
+);
 
-const tip = computed(() => {
-	if (!ndvStore.hasInputData) {
+const emptyExpression = computed(() => props.unresolvedExpression.trim().length === 0);
+
+const tip = computed<TipId>(() => {
+	if (!ndvStore.hasInputData && ndvStore.isInputParentOfActiveNode) {
 		return 'executePrevious';
 	}
 
-	if (props.tip) return props.tip;
+	if (canAddDotToExpression.value) {
+		return resolvedExpressionHasFields.value ? 'dotObject' : 'dotPrimitive';
+	}
 
-	if (ndvStore.focusedMappableInput) return 'drag';
+	if (canDragToFocusedInput.value && emptyExpression.value) return 'drag';
 
 	return 'default';
 });
-const expressionsDocsUrl = EXPRESSIONS_DOCS_URL;
+
+const cursor = computed(() => props.selection.anchor);
+
+function getCompletionsWithDot(): readonly Completion[] {
+	const atCursor = cursor.value;
+	if (
+		!props.editorState ||
+		!props.selection ||
+		!props.selection.empty ||
+		!props.unresolvedExpression ||
+		props.unresolvedExpression.charAt(atCursor - 1) === '.' ||
+		props.unresolvedExpression.charAt(atCursor - 1) === ' ' ||
+		props.unresolvedExpression.charAt(atCursor) === '.'
+	) {
+		return [];
+	}
+
+	const cursorAfterDot = atCursor + 1;
+	const docWithDot =
+		props.editorState.sliceDoc(0, atCursor) + '.' + props.editorState.sliceDoc(atCursor);
+	const selectionWithDot = EditorSelection.create([EditorSelection.cursor(cursorAfterDot)]);
+	const stateWithDot = EditorState.create({
+		doc: docWithDot,
+		selection: selectionWithDot,
+	});
+
+	const context = new CompletionContext(stateWithDot, cursorAfterDot, true);
+	const completionResult = datatypeCompletions(context);
+	return completionResult?.options ?? [];
+}
+
+watch(tip, (newTip) => {
+	ndvStore.setHighlightDraggables(!ndvStore.isMappingOnboarded && newTip === 'drag');
+});
+
+watchDebounced([() => props.selection, () => props.unresolvedExpression], () => {
+	const completions = getCompletionsWithDot();
+	canAddDotToExpression.value = completions.length > 0;
```

Comment:
Cases to consider:

1. Parens around objects and arrays e.g. `{{ ([1, 2, 3])| }}` and `{{ ({ a: 1 })| }}` do not trigger the message "Type `.` for data transformation options". This is valid access.

2. Oddly, no-parens access e.g. `{{ { a: 1 }| }}` → `{{ { a: 1 }.a }}` resolves despite not being valid JS. Not sure if we should support this because it is possible, or not support this because it is non-standard JS.

2. [Empty selection starting at dot position](https://share.cleanshot.com/xXLqJtl1) (still a valid position for access) does not trigger the message "Type `.` for data transformation options".

Reply:
Case 1 and 3 are fixed 🎉 

For 2, I'm not sure if we support this on purpose. But for the scope of this feature I don't think we should suggest adding a `.` in this case, wdyt?